### PR TITLE
Added follow symlink option to abstract watch service.

### LINKED ISSUE #2034: AbstractWatchService does not follow symlinks
**Problem**
The automatic reload of models fails if the folder to listen for changes is symlinked. For this reason ESH does not trigger reloads of items, rules etc.

**Steps to reproduce**
The easiest way to reproduce this is using openHAB. 
* Download openHAB 2.0 SNAPSHOT with demo items on a linux based machine and install it
* Copy the` [openhab-home]/conf` folder to another folder e.g` /opt/test/conf`
* Add `symlink ln -s /opt/test/conf [openhab-home]/conf`
* Start openHAB
* Change one of the models (e.g rule, items, sitemap,...)

 -> *The changed file will not be recognized. The log message 'refreshing MODEL' does not appear.*

**Cause**
The used AbstractWatchService does not follow symlinks. See: https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/AbstractWatchService.java#L80

For walking through files symlink should be followed. For this reason `FileVisitOption.FOLLOW_LINKS` should be added

### PR DESCRIPTION
**Pull request by chrisschauer**
Fixes #2034 

Signed-off-by: Christoph Schauer <christoph@schauer-itconsulting.de>

### CODE REVIEW

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/AbstractWatchService.java [hunk 1]
```diff
@@ -36,6 +39,11 @@
      * Default logger for ESH Watch Services
      */
     protected final Logger logger = LoggerFactory.getLogger(this.getClass());
+    
+    /**
+     * Tthe maximum number of directory levels to visit
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/AbstractWatchService.java]: typo

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/AbstractWatchService.java [hunk 2]
```diff
@@ -77,7 +85,9 @@ protected void initializeWatchService() {
                 if (watchSubDirectories()) {
                     watchService = FileSystems.getDefault().newWatchService();
 
-                    Files.walkFileTree(toWatch, new SimpleFileVisitor<Path>() {
+                    // walk through all folders and follow symlinks
+                    Files.walkFileTree(toWatch, Sets.newHashSet(FileVisitOption.FOLLOW_LINKS), MAX_DEPTH,
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/AbstractWatchService.java]: * How is the max depth related to "adding follow symlink option"?
* Why do we need to restrict the depth of the abstract watch service? Isn't it a specific usage if the depth is limited to a specific value?

Comment (chrisschauer) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/AbstractWatchService.java]: > How is the max depth related to "adding follow symlink option"?

IMHO it's not related, but it's a required parameter when using options: 
https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#walkFileTree-java.nio.file.Path-java.util.Set-int-java.nio.file.FileVisitor-

>Why do we need to restrict the depth of the abstract watch service? Isn't it a specific usage if the depth is limited to a specific value? 

O.K. good point. I thought on my case with only depth of 2 levels (conf/ and conf/rules, etc.). Maybe we should use `Integer.MAX_VALUE` so all levels should be visited. Preventing cycles is done by the Files implementations. 

Do you prefer using `Integer.MAX_VALUE `

Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/AbstractWatchService.java]: `Files.walkFileTree(toWatch, new SimpleFileVisitor<Path>() {...});` has been used before.

The JavaDoc you linked states:
```text
public static Path walkFileTree(Path start,
                                FileVisitor<? super Path> visitor)
                         throws IOException
Walks a file tree.
This method works as if invoking it were equivalent to evaluating the expression:

 walkFileTree(start, EnumSet.noneOf(FileVisitOption.class), Integer.MAX_VALUE, visitor)
 
In other words, it does not follow symbolic links, and visits all levels of the file tree.
```

So, if you want to follow symbolic links, too I would prefer to change only that parameter and keep the other defaults.

If we also want to be able to change the depth, we should do it in a separate PR.

> @maggu2810: Do you prefer using Integer.MAX_VALUE

Yes :wink:

**Comment by maggu2810** [comment]
Thanks
